test/integration/framework/master_utils.go

```diff
@@ -453,6 +454,25 @@ func ScaleRC(name, ns string, replicas int32, clientset internalclientset.Interf
 	return scaled, nil
 }
 
+// ScaleRS scales the given RS to the given replicas.
+func ScaleRS(name, ns string, replicas int32, clientset internalclientset.Interface) (*apiextensions.ReplicaSet, error) {
```

Comment:
@enisoc I looked into the codes for autoscaler repo, but found out that the HPA scaling tests are meant for the whole cluster, not a particular controller: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/scale_up_test.go

I found out that there is already a helper function for ReplicationController to test the scale endpoint: https://github.com/kubernetes/kubernetes/blob/master/test/integration/framework/master_utils.go#L438

The test uses internal clientset. I am able to write the integration test using the internal clientset, but it requires some redundant functions. It looks like `ReplicaSetScaler` also uses internal version APIs: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/scale.go#L262

I will continue finding the way to test against `extensions/v1beta1`.

Reply:
Be careful that there are multiple components called autoscaler. You found the Cluster Autoscaler but that is different from the Horizontal Pod Autoscaler. The CA adds or removes Nodes; the HPA adds or removes Pods.

The helper you linked for RC actually does not use the /scale endpoint (it does Get/Update on the full spec), so that's not the example we want to emulate.

To use the /scale endpoint with RS, it should be possible to just use the clientset you already have in the integration test and use the methods `clientset.ExtensionsV1beta1().Scales(namespace).Get(kind, name)` and `clientset.ExtensionsV1beta1().Scales(namespace).Update(kind, scale)`. Remember to use RetryOnConflict for all Get-edit-Update operations.

Reply:
Sound good to me! Let me try with the suggested methods.